src/main/java/org/jooq/lambda/Seq.java

```diff
@@ -9165,43 +9156,8 @@ public R next() {
      * </pre></code>
      */
     static <T> Tuple2<Seq<T>, Seq<T>> duplicate(Stream<? extends T> stream) {
-        final Iterator<? extends T> it = stream.iterator();
-        final LinkedList<T> gap = new LinkedList<>();
-
-        @SuppressWarnings("unchecked")
-        final Iterator<T>[] ahead = new Iterator[] { null };
-
-        class Duplicate implements Iterator<T> {
-            @Override
-            public boolean hasNext() {
-                if (ahead[0] == null || ahead[0] == this)
-                    return it.hasNext();
-
-                return !gap.isEmpty();
-            }
-
-            @Override
-            public T next() {
-                if (ahead[0] == null)
-                    ahead[0] = this;
-
-                if (ahead[0] == this) {
-                    T value = it.next();
-                    gap.offer(value);
-                    return value;
-                }
-                else {
-                    T value = gap.poll();
-
-                    if (gap.isEmpty())
-                        ahead[0] = null;
-
-                    return value;
-                }
-            }
-        }
-
-        return tuple(seq(new Duplicate()), seq(new Duplicate()));
+        SeqBuffer<T> buffer = SeqBuffer.of(stream);
+        return tuple(buffer.seq(), buffer.seq());
```

Comment:
You're absolutely right. On the other hand, current implementation is thread safe thanks to `SeqBuffer`, and the previous one wasn't.

All in all, I'd be more inclined to this simple implementation but I guess I'd leave an appropriate comment in this method in case someone reports any memory-related issues in the future.